addons/misra.py

```diff
@@ -1937,6 +1937,38 @@ def misra_20_14(self, data):
                     ifStack.pop()
 
 
+    def misra_21_1(self, data):
+        # Reference: n1570 7.1.3 - Reserved identifiers
+        re_forbidden_macro = re.compile(r'#define (errno|_[_A-Z]+)')
+
+        # Search for forbidden identifiers in macro names
+        for directive in data.directives:
+            res = re.search(re_forbidden_macro, directive.str)
+            if res:
+                self.reportError(directive, 21, 1)
+
+        # Search for forbidden identifiers
+        for token in data.tokenlist:
+            if not token.isName:
+                continue
+            if len(token.str) < 2:
+                continue
+            if token.str == 'errno':
+                self.reportError(token, 21, 1)
+            if token.str[0] == '_':
+                if (token.str[1] in string.ascii_uppercase) or (token.str[1] == '_'):
```

Comment:
hmm.. to me it sounds slow to use `token.str[1] in string.ascii_uppercase`.. does that take 25 comparisons to determine that it's not uppercase?

Reply:
Hm, as C programmer the most oblivious "optimization" for me would be using comparison with symbol `ord`. But interpreter doesn't know about `char` type and it will compare `PyObject`'s anyway, so it would be even slower.

Consider example:
```python
import time
import random
import string
import sys

def randomString(stringLength=10):
    """Generate a random string of fixed length """
    letters = string.ascii_lowercase
    return ''.join(random.choice(letters) for i in range(stringLength))

def s1(c):
    return c in string.ascii_uppercase

def s2(c):
    return ord('A') < c < ord('Z')

if __name__ == '__main__':
    s = randomString(100000)
    start_time = None
    if sys.argv[1] == '1':
        start_time = time.time()
        for c in s:
            s1(c)
    else:
        start_time = time.time()
        for c in s:
            s2(c)
    print("--- %s seconds ---" % (time.time() - start_time))
``` 

Results:
```bash
/tmp python 1.py 1
--- 0.0154910087585 seconds ---
/tmp python 1.py 1
--- 0.0146980285645 seconds ---
/tmp python 1.py 1
--- 0.0144381523132 seconds ---
/tmp python 1.py 1
--- 0.014445066452 seconds ---
/tmp python 1.py 1
--- 0.0146498680115 seconds ---
/tmp python 1.py 2 
--- 0.0261898040771 seconds ---
/tmp python 1.py 2
--- 0.0264759063721 seconds ---
/tmp python 1.py 2
--- 0.0262699127197 seconds ---   
```

Reply:
It also interesting to check `dis.dis` output for these functions:
```
In [5]: dis.dis(s1)
  2           0 LOAD_FAST                0 (c)
              3 LOAD_GLOBAL              0 (string)
              6 LOAD_ATTR                1 (ascii_upper)
              9 COMPARE_OP               6 (in)
             12 RETURN_VALUE    
```

```
In [5]: dis.dis(s1)
  2           0 LOAD_FAST                0 (c)
              3 LOAD_GLOBAL              0 (string)
              6 LOAD_ATTR                1 (ascii_upper)
              9 COMPARE_OP               6 (in)
             12 RETURN_VALUE    
```